runtime/registry.go

```diff
@@ -187,35 +195,88 @@ func (r *Runtime) EditInstance(ctx context.Context, inst *drivers.Instance) erro
 // TODO :: this is a rudimentary solution and ideally should be done in some producer/consumer pattern
 func (r *Runtime) evictCaches(ctx context.Context, inst *drivers.Instance) {
 	// evict and close exisiting connection
-	r.connCache.evict(ctx, inst.ID, inst.OLAPDriver, inst.OLAPDSN)
-	r.connCache.evict(ctx, inst.ID, inst.RepoDriver, inst.RepoDSN)
+	c, _, _ := r.OLAPDef(inst)
+	r.connCache.evict(ctx, inst.ID, c.Type, r.variables(inst.OLAPDriver, c.Configs, inst.ResolveVariables()))
+	c, _, _ = r.RepoDef(inst)
+	r.connCache.evict(ctx, inst.ID, c.Type, r.variables(inst.RepoDriver, c.Configs, inst.ResolveVariables()))
 
 	// evict catalog cache
 	r.migrationMetaCache.evict(ctx, inst.ID)
 	// query cache can't be evicted since key is a combination of instance ID and other parameters
 }
 
-func (r *Runtime) checkRepoConnection(inst *drivers.Instance) (drivers.Connection, drivers.RepoStore, error) {
-	repo, err := drivers.Open(inst.RepoDriver, map[string]any{"dsn": inst.RepoDSN}, r.logger)
+func (r *Runtime) checkRepoConnection(inst *drivers.Instance) (drivers.Handle, drivers.RepoStore, error) {
+	c, shared, err := r.RepoDef(inst)
 	if err != nil {
 		return nil, nil, err
 	}
-	repoStore, ok := repo.AsRepoStore()
+	repo, err := drivers.Open(c.Type, r.variables(c.Name, c.Configs, inst.ResolveVariables()), shared, r.logger)
+	if err != nil {
+		return nil, nil, err
+	}
+	repoStore, ok := repo.AsRepoStore(inst.ID)
 	if !ok {
 		return nil, nil, fmt.Errorf("not a valid repo driver: '%s'", inst.RepoDriver)
 	}
 
 	return repo, repoStore, nil
 }
 
-func (r *Runtime) checkOlapConnection(inst *drivers.Instance) (drivers.Connection, drivers.OLAPStore, error) {
-	olap, err := drivers.Open(inst.OLAPDriver, map[string]any{"dsn": inst.OLAPDSN}, r.logger)
+func (r *Runtime) checkOlapConnection(inst *drivers.Instance) (drivers.Handle, drivers.OLAPStore, error) {
+	c, shared, err := r.OLAPDef(inst)
+	if err != nil {
+		return nil, nil, err
+	}
+	olap, err := drivers.Open(c.Type, r.variables(c.Name, c.Configs, inst.ResolveVariables()), shared, r.logger)
 	if err != nil {
 		return nil, nil, err
 	}
-	olapStore, ok := olap.AsOLAP()
+	olapStore, ok := olap.AsOLAP(inst.ID)
 	if !ok {
 		return nil, nil, fmt.Errorf("not a valid OLAP driver: '%s'", inst.OLAPDriver)
 	}
 	return olap, olapStore, nil
 }
```

Comment:
Could we leverage some generic functions here? It doesn't feel great to be crafting these repo and OLAP `Open` payloads multiple places

Reply:
Responded here : https://github.com/rilldata/rill/pull/2875#discussion_r1291019218